### DISCUSSION TITLE
How do I generate a validation curve with LSTMFCNClassifier?

### DISCUSSION THREAD

**Discussion by TannerCypret** [alan-turing-institute/sktime#7944] [Q&A]
I am creating the LSTMFCN classifier like this and including a tensorboard call back: 

`lstm = LSTMFCNCLassifier(n_epochs=700,batch_size=16,callbacks=[tf.keras.callbacks.EarlyStopping('accuracy',patience=25),tensorboard_callback],attention=False)`

`lstm.fit(x_train,y_train)`

When I use the fit method Tensorboard creates a nice graph of accuracy over epochs during training. I would like a validation curve to go with it. Comparing the training vs validation curve would help me detect issues like overfitting.

Looking at the [documentation ](https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.classification.deep_learning.LSTMFCNClassifier.html)I found the fit_predict function to call that can take a cross validation object; however, I am struggling to get it to work. Currently my input data is in the pd-multiindex dataframe format. I am trying to train with a validation set like this:

`kf = kFold(shuffle=True, random_state=100)`
`kf.split(X=x_val,y=y_val)`

`lstm = LSTMFCNCLassifier(n_epochs=700,batch_size=16,callbacks=[tf.keras.callbacks.EarlyStopping('accuracy',patience=25),tensorboard_callback],attention=False)`

`lstm.fit_predict(x_train,y_train,cv=kf,changeState=True)`

This gives me the following error:

`numpy.core._exceptions._ufunc 'negative' did not contain a loop with signature matching types <class 'numpy.dtypes.strDType'> -> None`

I think the issue is that my kf object is expecting numpy arrays and I have training data in the form of a dataframe. Is there another way to do this? The data in the cross validation object should be the same data type as the other training data, which for me is a pd-multiindex dataframe time series. Can anyone provide any examples of how to call fit_predict with validation data such that you get a validation curve generated by tensorabord?

**Comment by TannerCypret** [comment]
I checked the model summary and it has accuracy and loss but it doesn't have separate curves for test/train accuracy and loss. Is there a way to store data for both?

It looks like for sklearn you can set warm_start=true and then repeatedly call fit in a loop set to the number of epochs. You could then call the score method after each epoch and store that value to see test performance as the model is iteratively trained. LSTM-FCN does not support the warm_start parameter. So it looks like this solution won't work either.

**Comment by arnavk23** [comment]
Hi @TannerCypret,

Great question—getting validation curves from `LSTMFCNClassifier` in `sktime` can be a bit tricky since it's built on top of `sktime-dl`, which uses Keras under the hood, and doesn’t directly expose the full training history like a typical Keras model.

You're right that:

* `fit_predict(..., cv=...)` is primarily for evaluation, not training with validation splits per epoch.
* Your current cross-validator (`KFold`) expects NumPy arrays, but `LSTMFCNClassifier` expects `Panel`-type input (like a MultiIndex `DataFrame`).
* There’s no built-in `warm_start` mechanism or per-epoch external validation scoring in the current `LSTMFCNClassifier` interface.


### Suggested Solution: Use `validation_data` in a Keras Callback

To generate training vs validation curves for **each epoch**, you can pass a `tf.keras.callbacks.Callback` like `TensorBoard` or `History`, but you need to ensure that the model gets a **validation set** during training.

Although `LSTMFCNClassifier.fit()` doesn’t expose `validation_data` directly, it **does** accept `fit_kwargs`, which get passed to the underlying `model.fit()`.

So you can do:

```python
# Split training and validation manually
x_train_split, x_val_split = x_train.iloc[train_index], x_train.iloc[val_index]
y_train_split, y_val_split = y_train.iloc[train_index], y_train.iloc[val_index]

# Define model
lstm = LSTMFCNClassifier(
    n_epochs=700,
    batch_size=16,
    callbacks=[
        tf.keras.callbacks.EarlyStopping('val_accuracy', patience=25),
        tensorboard_callback
    ],
    attention=False
)

# Fit with explicit validation data
lstm.fit(
    x_train_split,
    y_train_split,
    fit_kwargs={
        "validation_data": (x_val_split, y_val_split)
    }
)
```

This way, `TensorBoard` will log both training and validation metrics per epoch. You’ll get separate curves for `accuracy` and `val_accuracy`, and the same for `loss`.

### Notes on Cross-Validation

If you want to do proper cross-validation *and* log validation curves:

* You’ll need to **manually loop over folds**, and train one model per fold.
* Collect each fold’s history if needed.

Example (simplified):

```python
from sklearn.model_selection import KFold

kf = KFold(n_splits=3, shuffle=True, random_state=42)
for fold, (train_idx, val_idx) in enumerate(kf.split(x_train)):
    x_tr, x_val = x_train.iloc[train_idx], x_train.iloc[val_idx]
    y_tr, y_val = y_train.iloc[train_idx], y_train.iloc[val_idx]

    lstm = LSTMFCNClassifier(...)  # same as above

    lstm.fit(x_tr, y_tr, fit_kwargs={"validation_data": (x_val, y_val)})
```

### Summary

* Use `fit_kwargs={"validation_data": ...}` to log per-epoch validation performance with TensorBoard.
* Cross-validation + validation curves requires manual looping.
* `warm_start` and `partial_fit` aren't currently supported by LSTM-FCN in `sktime`.
